### ISSUE TITLE
Missing value not skipped by SkipMissingVal

### ISSUE THREAD

**Issue by maxfreu** [rafaqz/Rasters.jl#245]
I encountered a really strange bug where the no data value is not skipped by skipmissing. Seems like the cause is that the missingval is Int16 and the array eltype is Float32. You can download the data [here](https://owncloud.gwdg.de/index.php/s/AWBgUGw7Zw2K6gS) and [here](https://owncloud.gwdg.de/index.php/s/sxZKTb2MpnxghND).

```julia
foo = read(Raster("missingnotskipped.tif"))
bar = deserialize("missingnotskipped.jls")
qwe = rebuild(bar, missingval=-999f0)

all(foo .== bar .== qwe)  # true
typeof(foo) == typeof(bar)  # false

missingval(foo) in Rasters.SkipMissingVal(foo)  # false
missingval(bar) in Rasters.SkipMissingVal(bar)  # true
missingval(qwe) in Rasters.SkipMissingVal(qwe)  # false

skipmissing_(r) = (v for v in r if v != missingval(r))

missingval(bar) in skipmissing_(bar)  # false
```

Edit: Could it be that [this line](https://github.com/rafaqz/Rasters.jl/blob/1f1bcb64972ee774746e515fcf3370b0ca91f6fc/src/skipmissing.jl#L32) should have a `==` for the comparison with the missingval?

**Comment by rafaqz** [comment]
Yeah the `missingval` should be the same type... because `missing` needs `===` equality. 

We could special case `missing` and use `==` for everything else. Otherwise we could force the type with `promote_type` when the tif is loaded. 

Honestly I think its a bug in GDAL that the nodata value can be a different type to the array. It makes no sense.

**Comment by maxfreu** [comment]
To clafiy: the discrepancy originated not upon loading, but when I converted some arbitrary raster from Int16 to Float32 via `float32.(raster)` from `ImageCore` (or float or any other type conversion). The demo raster I uploaded shows that it works ok once I write it to disk and read it back in; I think GDAL is fine. But yes, using `==` in the line quoted above would be good to handle situations where the types differ. In real world use value equality should be enough. The other way to go would be to ensure that all type conversions also affect the missingval.
